src/core/mesh/qgsmesh3daveraging.cpp

```diff
@@ -151,3 +248,513 @@ int QgsMeshSingleLevelAveragingMethod::verticalLevel() const
+{
+  return new QgsMeshHeightAveragingMethod( startHeight(), endHeight() );
+}
+
+double QgsMeshHeightAveragingMethod::startHeight() const
+{
+  return mStartHeight;
+}
+
+double QgsMeshHeightAveragingMethod::endHeight() const
+{
+  return mEndHeight;
+}
+
+bool QgsMeshHeightAveragingMethod::hasValidInputs() const
+{
+  return mStartHeight >= 0 && mEndHeight >= mStartHeight;
+}
+
+void QgsMeshHeightAveragingMethod::volumeRangeForFace(
+  double &startVerticalLevel,
+  double &endVerticalLevel,
+  const QVector<double> verticalLevels ) const
+{
+  const double bot = verticalLevels[verticalLevels.size() - 1];
+  startVerticalLevel = bot + mStartHeight;
+  endVerticalLevel = bot + mEndHeight;
+}
+
+QgsMeshElevationAveragingMethod::QgsMeshElevationAveragingMethod()
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::ElevationAveragingMethod )
+{
+}
+
+QgsMeshElevationAveragingMethod::QgsMeshElevationAveragingMethod( double startElevation, double endElevation )
+  : QgsMesh3dAveragingMethod( QgsMesh3dAveragingMethod::ElevationAveragingMethod )
+  , mStartElevation( startElevation )
+  , mEndElevation( endElevation )
+{
+  if ( mEndElevation > mStartElevation )
+  {
+    std::swap( mEndElevation, mStartElevation );
+  }
+}
+
+QgsMeshElevationAveragingMethod::~QgsMeshElevationAveragingMethod() = default;
+
+QDomElement QgsMeshElevationAveragingMethod::writeXml( QDomDocument &doc ) const
+{
+  QDomElement elem = doc.createElement( QStringLiteral( "elevation-settings" ) );
+  elem.setAttribute( QStringLiteral( "start-elevation" ), startElevation() );
+  elem.setAttribute( QStringLiteral( "end-elevation" ), endElevation() );
+  return elem;
+}
+
+void QgsMeshElevationAveragingMethod::readXml( const QDomElement &elem )
+{
+  mStartElevation = elem.attribute( QStringLiteral( "start-elevation" ) ).toInt();
+  mEndElevation = elem.attribute( QStringLiteral( "end-elevation" ) ).toInt();
+  if ( mEndElevation > mStartElevation )
+  {
+    std::swap( mEndElevation, mStartElevation );
+  }
+}
+
+bool QgsMeshElevationAveragingMethod::equals( const QgsMesh3dAveragingMethod *other ) const
+{
+  if ( !other || other->method() != method() )
+    return false;
+
+  const QgsMeshElevationAveragingMethod *otherMethod = static_cast<const QgsMeshElevationAveragingMethod *>( other );
+
+  return qgsDoubleNear( otherMethod->startElevation(), startElevation() ) && qgsDoubleNear( otherMethod->endElevation(), endElevation() ) ;
+}
+
+QgsMesh3dAveragingMethod *QgsMeshElevationAveragingMethod::clone() const
+{
+  return new QgsMeshElevationAveragingMethod( startElevation(), endElevation() );
+}
+
+double QgsMeshElevationAveragingMethod::startElevation() const
+{
+  return mStartElevation;
+}
+
+double QgsMeshElevationAveragingMethod::endElevation() const
+{
+  return mEndElevation;
+}
+
+bool QgsMeshElevationAveragingMethod::hasValidInputs() const
+{
+  return mStartElevation <= 0.0 && mEndElevation <= mStartElevation;
```

Comment:
I'm not sure if I understand why both of these checks are necessary?

(it is counter-intuitive that end elevation should not be more than start elevation - normally start <= end)

Reply:
we are counting from 0 (surface) down (so the elevation is based on model absolute datum)